### ISSUE TITLE
change replicas aborted due to failed preemptive snapshot ... cannot apply snapshot: snapshot intersects existing range; initiated GC

### ISSUE THREAD

**Issue by youngl98** [cockroachdb/cockroach#13042]
I have a 3 node cluster running the 20170112 binaries on all 3 nodes.  The cluster crashed 3 days ago, when the cluster was restarted.  We are seeing these kinds of messages in the ERROR log files on all 3 nodes.
```
E170121 18:30:18.929514 31 storage/queue.go:599  [raftsnapshot,n1,s1,r845/3:/System/tsd/cr.store.replicas…,@c4200a2900] [n1,s1,r845/3:/System/tsd/cr.store.replicas…]: change replicas aborted due to failed preemptive snapshot: range=845: remote couldn't accept snapshot with error: [n2,s2],r845: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n2,s2,r347/1:/System/tsd/cr.store.replicas…]
```
The cluster seems to be in some sort or a replica GC loop.  What should one do to fix this?  Or will cockroach eventually repair by itself?

Attaching log files from all 3 nodes for
[node1-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/721390/node1-logs.tar.gz)
[node2-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/721391/node2-logs.tar.gz)
[node3-logs.tar.gz](https://github.com/cockroachdb/cockroach/files/721392/node3-logs.tar.gz)

 review.

**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

     grep -F '[config]' cockroach-data/logs/cockroach.INFO
```
I170121 17:31:19.771189 1 util/log/clog.go:1003  [config] file created at: 2017/01/21 17:31:19
I170121 17:31:19.771189 1 util/log/clog.go:1003  [config] running on machine: XXXXXXXXXX
I170121 17:31:19.771189 1 util/log/clog.go:1003  [config] binary: CockroachDB beta-20170112 (linux amd64, built 2017/01/12 18:27:36, go1.7.3)
I170121 17:31:19.771189 1 util/log/clog.go:1003  [config] arguments: [/var/cockroach/bin/cockroach start --http-port=8080 --store=path=/var/cockroach/cockroach-data --store=path=/usr/mware/cockroachdb-store --key=/var/cockroach/certs/XXXXXX.key]
```

**Comment by bdarnell** [comment]
Aside from these log messages, is the cluster working? It's normal to see some of these messages during the recovery process, but they should clear up on their own relatively quickly (minutes or maybe hours; certainly not days). 

We're currently investigating some problems that could cause recovery to get stuck, but our current understanding of the issue is that it's caused by code that is not enabled by default in the 0112 beta.

**Comment by youngl98** [comment]
I am attaching the cockroach console screenshots.  They appear to be ok.

I can perform queries against the cluster, so on the surface it appears to be working fine.

I am still getting these messages in the ERROR log
```
E170123 17:51:16.924346 31 storage/queue.go:599  [raftsnapshot,n1,s1,r845/3:/System/tsd/cr.store.replicas…,@c4200a2900] [n1,s1,r845/3:/System/tsd/cr.store.replicas…]: change replicas aborted due to failed preemptive snapshot: range=845: remote couldn't accept snapshot with error: [n2,s2],r845: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n2,s2,r347/1:/System/tsd/cr.store.replicas…]
E170123 17:51:17.923798 31 storage/queue.go:599  [raftsnapshot,n1,s1,r845/3:/System/tsd/cr.store.replicas…,@c4200a2900] [n1,s1,r845/3:/System/tsd/cr.store.replicas…]: change replicas aborted due to failed preemptive snapshot: range=845: remote couldn't accept snapshot with error: [n2,s2],r845: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n2,s2,r347/1:/System/tsd/cr.store.replicas…]
E170123 17:51:18.926097 31 storage/queue.go:599  [raftsnapshot,n1,s1,r845/3:/System/tsd/cr.store.replicas…,@c4200a2900] [n1,s1,r845/3:/System/tsd/cr.store.replicas…]: change replicas aborted due to failed preemptive snapshot: range=845: remote couldn't accept snapshot with error: [n2,s2],r845: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n2,s2,r347/1:/System/tsd/cr.store.replicas…]
```
![screen shot 2017-01-23 at 9 50 35 am](https://cloud.githubusercontent.com/assets/322059/22215950/fa938e54-e151-11e6-879f-6bf3cf8d9902.jpg)
![screen shot 2017-01-23 at 9 50 40 am](https://cloud.githubusercontent.com/assets/322059/22215951/faca15aa-e151-11e6-893a-d51e1d1922c9.jpg)

**Comment by cuongdo** [comment]
Last's night's 4-node, 6-hour `block_writer` test failed, because no queries were served for 2 minutes. [Full logs here](https://teamcity.cockroachdb.com/viewLog.html?buildId=140522&buildTypeId=C_BW_4Nodes_6Hours&tab=artifacts)

[node.0/cockroach.stderr](https://teamcity.cockroachdb.com/repository/download/C_BW_4Nodes_6Hours/140522:id/node.0/cockroach.stderr) also reveals a similar looking issue:

```
E170206 07:36:09.249376 64 storage/queue.go:628  [raftsnapshot,n1,s1,r14/1:/Table/51/1/{577163…-634609…},@c420acaf00] [n1,s1,r14/1:/Table/51/1/{577163…-634609…}]: change replicas aborted due to failed preemptive snapshot: r14: remote couldn't accept snapshot with error: [n4,s4],r14: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n4,s4,r9/2:/Table/51/1/{461836…-692280…}]
```

[node.1/cockroach.stderr](https://teamcity.cockroachdb.com/repository/download/C_BW_4Nodes_6Hours/140522:id/node.1/cockroach.stderr) shows many of the following messages as well:
```
I170206 07:35:33.801820 125 storage/replicate_queue.go:228  [replicate,n2,s2,r3/2:/Table/1{1-2},@c4203a2900] [n2,s2,r3/2:/Table/1{1-2}]: change replicas aborted due to failed preemptive snapshot: r3: remote declined snapshot: reservation rejected
E170206 07:35:42.894050 125 storage/queue.go:628  [replicate,n2,s2,r3/2:/Table/1{1-2},@c4203a2900] failed to replicate [n2,s2,r3/2:/Table/1{1-2}] after 5 retries
```

which was also accompanied by these messages:

```
W170206 07:36:42.541226 811 kv/dist_sender.go:1166  [client=127.0.0.1:42648,user=root,n1] have been waiting 1m0s sending RPC for batch: 1 CPut, 1 BeginTxn, 1 EndTxn
```


[node.4/cockroach.stderr](https://teamcity.cockroachdb.com/repository/download/C_BW_4Nodes_6Hours/140522:id/node.3/cockroach.stderr) shows a series of these messages:

```
W170206 07:36:58.584151 897641 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.584159 897597 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.584184 897575 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.584192 897579 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.584221 897617 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.584226 897640 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280…-807042…},@c420950300] have been waiting 1m0s attempting to acquire lease
```

When the test terminated, all `block_writer` instances had 0 qps.

**Comment by bdarnell** [comment]
In those logs the "snapshot intersecting existing range" issue is affecting a lot of ranges, but let's focus on one. At 07:24:26, node 0 sends a snapshot of range 12 (covering keys `/Table/51/1/{115490...-231054...}`) to node 3, which rejects it because it conflicts with that node's replica of range 8 (covering keys `/Table/51{-/1/23105â€¦}`).

```
E170206 07:24:26.137465 64 storage/queue.go:628  [raftsnapshot,n1,s1,r12/1:/Table/51/1/{115490â€¦-231054â€¦},@c4211bc300] [n1,s1,r12/1:/Table/51/1/{115490â€¦-231054â€¦}]: change replicas aborted due to failed preemptive snapshot: r12: remote couldn't accept snapshot with error: [n4,s4],r12: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n4,s4,r8/2:/Table/51{-/1/23105â€¦}]
```

So range 8 split at some point, and range 12 is its new RHS. That split happened 6 seconds earlier: 

```
I170206 07:24:20.931745 60 storage/replica_command.go:2372  [split,n1,s1,r8/1:/Table/51{-/1/23105â€¦},@c420de5800] initiating a split of this range at key /Table/51/1/1154905455289310174/"f6a852ee-d93c-44ed-b8a6-4bed6d0b567a"/38016 [r12]
```

Node 3 is falling behind badly: It's applying a lot of raft snapshots, and seeing a lot of commits that are taking longer than 500ms (up to several seconds). 

```
I170206 07:24:29.815664 440745 storage/replica_raftstorage.go:594  [n4,s4,r8/2:/Table/51{-/1/23105â€¦},@c4201edb00] applying Raft snapshot at index 956403 (id=87fe233f, encoded size=34650067, 33 rocksdb batches, 851 log entries)
...
I170206 07:24:33.354022 440745 storage/replica_raftstorage.go:602  [n4,s4,r8/2:/Table/51{-/1/11549â€¦},@c4201edb00] applied Raft snapshot in 3538ms [clear=558ms batch=114ms entries=2ms commit=2863ms]
```

In these two lines we see that node 3 finally learned about the split 13 seconds after it happened, after it had already rejected a preemptive snapshot. After this, we no longer see "intersects existing range" errors for `r8`.

Here's an especially egregious example of the same pattern. Node 4 fell *12 minutes* behind on range 9 and missed out on two separate splits:

Node 0: 
```
I170206 07:24:24.396896 60 storage/replica_command.go:2372  [split,n1,s1,r9/1:/Table/51/1/{461836â€¦-692280â€¦},@c4222ff500] initiating a split of this range at key /Table/51/1/5771636842022332685/"5dd77265-a2e9-4f2f-9e2c-a16c8b0d80bd"/53781 [r14]
...
I170206 07:35:37.484498 60 storage/replica_command.go:2372  [split,n1,s1,r9/1:/Table/51/1/{461836â€¦-577163â€¦},@c4222ff500] initiating a split of this range at key /Table/51/1/5193947981976960133/"4e58c08f-c77a-48c0-90ed-bf5da29f4d08"/3743 [r21]
```

Node 3: 
```
I170206 07:36:10.498287 898368 storage/replica_raftstorage.go:594  [n4,s4,r9/2:/Table/51/1/{461836â€¦-692280â€¦},@c420d34000] applying Raft snapshot at index 713824 (id=ad4303bf, encoded size=34618065, 33 rocksdb batches, 696 log entries)
I170206 07:36:10.995899 898368 storage/replica_raftstorage.go:602  [n4,s4,r9/2:/Table/51/1/{461836â€¦-519394â€¦},@c420d34000] applied Raft snapshot in 498ms [clear=96ms batch=99ms entries=2ms commit=300ms]
```

Node 3 has clearly gone into a death spiral in which it falls behind, needs raft snapshots, and applying those snapshots is expensive enough that it falls further behind. Here's an example of it needing back-to-back snapshots:

```
I170206 07:18:15.930086 222697 storage/replica_raftstorage.go:594  [n4,s4,r10/2:/Table/51/1/{231054â€¦-461836â€¦},@c4209f7b00] applying Raft snapshot at index 3413 (id=983d7af1, encoded size=34660496, 33 rocksdb batches, 394 log entries)
...
I170206 07:18:17.004414 222697 storage/replica_raftstorage.go:602  [n4,s4,r10/2:/Table/51/1/{231054â€¦-461836â€¦},@c4209f7b00] applied Raft snapshot in 1074ms [clear=414ms batch=108ms entries=1ms commit=550ms]
I170206 07:18:18.946487 224467 storage/replica_raftstorage.go:594  [n4,s4,r10/2:/Table/51/1/{231054â€¦-461836â€¦},@c4209f7b00] applying Raft snapshot at index 5357 (id=71d73673, encoded size=35229538, 34 rocksdb batches, 830 log entries)
```

It is unclear whether there is any other cause or catalyst behind node 3's trouble or if it's just a case of it getting unlucky and falling far enough behind that it can't catch up (in which case some sort of flow control to slow down the other nodes would be the only solution I can see). It's also unclear how this turns into zero QPS since the other nodes are clearly able to make progress.

**Comment by bdarnell** [comment]
The "waiting 1m0s" messages in the node 3 log are all about range 11, and start at 07:36:42.

```
W170206 07:36:42.504203 349 kv/dist_sender.go:1166  [client=127.0.0.1:46830,user=root,n4] have been waiting 1m0s sending RPC for batch: 1 CPut, 1 BeginTxn, 1 EndTxn
W170206 07:36:42.523768 348 kv/dist_sender.go:1166  [client=127.0.0.1:46834,user=root,n4] have been waiting 1m0s sending RPC for batch: 1 CPut, 1 BeginTxn, 1 EndTxn
```

And then a second wave of slightly different messages at 07:36:58:

```
W170206 07:36:58.583856 897625 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] have been waiting 1m0s attempting to acquire lease
W170206 07:36:58.583888 897632 storage/replica.go:1101  [n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] have been waiting 1m0s attempting to acquire lease
```

What was going on with that range a minute earlier? On node 0:

```
I170206 07:35:39.732732 60 storage/split_queue.go:118  [split,n1,s1,r11/1:/Table/51/1/{692280â€¦-807042â€¦},@c423243500] splitting size=67109229 max=67108864
...
I170206 07:35:40.266513 60 storage/replica_command.go:2372  [split,n1,s1,r11/1:/Table/51/1/{692280â€¦-807042â€¦},@c423243500] initiating a split of this range at key /Table/51/1/7495503318271109140/"a30c23c9-ac96-4964-9d06-c9afcf9b5368"/76025 [r22]
I170206 07:35:40.391295 61 storage/replica_raftstorage.go:412  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-807042â€¦},@c423243500] generated preemptive snapshot 3a65a0f7 at index 474280
I170206 07:35:40.393586 61 storage/replicate_queue.go:228  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-807042â€¦},@c423243500] [n1,s1,r11/1:/Table/51/1/{692280â€¦-807042â€¦}]: change replicas aborted due to failed preemptive snapshot: r11: remote declined snapshot: reservation rejected
I170206 07:35:40.449906 61 storage/replica_raftstorage.go:412  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] generated preemptive snapshot 21346718 at index 474283
I170206 07:35:40.451873 61 storage/replicate_queue.go:228  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] [n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦}]: change replicas aborted due to failed preemptive snapshot: r11: remote declined snapshot: reservation rejected
I170206 07:35:40.544571 61 storage/replica_raftstorage.go:412  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] generated preemptive snapshot 965a51ce at index 474312
...
I170206 07:35:42.143929 61 storage/store.go:3266  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] streamed snapshot: kv pairs: 236410, log entries: 1424, 1599ms
I170206 07:35:42.379240 61 storage/replica_command.go:3228  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] change replicas (remove {2 2 4}): read existing descriptor range_id:11 start_key:"\273\211\375`\022\300\2275;\327\336\0220f09da66-53ca-4ae2-a552-c3df13c01ca2\000\001\367_:" end_key:"\273\211\375h\005`\216\032O\250\024\022a30c23c9-ac96-4964-9d06-c9afcf9b5368\000\001\370\001(\371" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:4 store_id:4 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:3 > next_replica_id:4 
I170206 07:35:42.408241 3646814 storage/replica.go:2465  [n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:4}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:4 StoreID:4 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3} {NodeID:2 StoreID:2 ReplicaID:4}]
```

Meanwhile, on node 1:
```
I170206 07:35:42.118269 940758 storage/replica_raftstorage.go:594  [n2,s2,r11/?:{-},@c4248c9800] applying preemptive snapshot at index 474312 (id=965a51ce, encoded size=35054126, 33 rocksdb batches, 1424 log entries)
...
I170206 07:35:42.349098 940758 storage/replica_raftstorage.go:602  [n2,s2,r11/?:/Table/51/1/{692280â€¦-749550â€¦},@c4248c9800] applied preemptive snapshot in 230ms [clear=0ms batch=114ms entries=4ms commit=111ms]
...
I170206 07:35:50.123453 128 storage/replica_raftstorage.go:412  [raftsnapshot,n2,s2,r11/4:/Table/51/1/{692280â€¦-749550â€¦},@c4248c9800] generated Raft snapshot 6cb0608d at index 474426
...
I170206 07:35:56.885918 128 storage/store.go:3266  [raftsnapshot,n2,s2,r11/4:/Table/51/1/{692280â€¦-749550â€¦},@c4248c9800] streamed snapshot: kv pairs: 236512, log entries: 115, 6593ms
```

Back to node 0:

```
I170206 07:35:54.887969 61 storage/replica_command.go:3228  [replicate,n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] change replicas (remove {4 4 2}): read existing descriptor range_id:11 start_key:"\273\211\375`\022\300\2275;\327\336\0220f09da66-53ca-4ae2-a552-c3df13c01ca2\000\001\367_:" end_key:"\273\211\375h\005`\216\032O\250\024\022a30c23c9-ac96-4964-9d06-c9afcf9b5368\000\001\370\001(\371" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:4 store_id:4 replica_id:2 > replicas:<node_id:3 store_id:3 replica_id:3 > replicas:<node_id:2 store_id:2 replica_id:4 > next_replica_id:5 
I170206 07:35:54.906781 3647550 storage/replica.go:2465  [n1,s1,r11/1:/Table/51/1/{692280â€¦-749550â€¦},@c423243500] proposing REMOVE_REPLICA {NodeID:4 StoreID:4 ReplicaID:2}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:4} {NodeID:3 StoreID:3 ReplicaID:3}]
```

So range 11 splits, and almost immediately does a ChangeReplicas to add node 1, and after a few rejected reservations it succeeds. The range is now replicated across all four nodes. Node 1 becomes the new leader of the range (based on the fact that it sends a raft snapshot), and it removes node 3. 

Node 3 sees its removal from the range and GCs its replica:

```
I170206 07:35:57.847281 210 storage/store.go:3122  [n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] added to replica GC queue (peer suggestion)
I170206 07:35:57.849271 210 storage/store.go:3122  [n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] added to replica GC queue (peer suggestion)
...
I170206 07:35:58.583413 171 storage/store.go:2092  [replicaGC,n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] removing replica
...
I170206 07:35:59.402688 171 storage/replica.go:736  [replicaGC,n4,s4,r11/2:/Table/51/1/{692280â€¦-807042â€¦},@c420950300] removed 381488 (381479+9) keys in 818ms [clear=371ms commit=447ms]
```

So it looks like the commands that were waiting to acquire the lease were not canceled as they should have been when the replica was removed at 7:35:58. I'm not sure why the commands at `:42` got stuck as well.

**Comment by bdarnell** [comment]
I think it was more likely fixed by #13515 (which would have been released in beta-20170216). I'm going to close this, but @youngl98 please let us know if you're still seeing this problem.
